src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,160 @@
+#include "distributed/metadata_utility.h"
+#include "distributed/multi_logical_replication.h"
+#include "distributed/multi_progress.h"
+#include "distributed/multi_server_executor.h"
+#include "distributed/pg_dist_rebalance_strategy.h"
+#include "distributed/pg_dist_shard.h"
+#include "distributed/reference_table_utils.h"
+#include "distributed/remote_commands.h"
+#include "distributed/resource_lock.h"
+#include "distributed/shard_rebalancer.h"
+#include "distributed/shard_cleaner.h"
+#include "distributed/shard_transfer.h"
+#include "distributed/tuplestore.h"
+#include "distributed/utils/array_type.h"
+#include "distributed/worker_protocol.h"
+#include "nodes/pg_list.h"
+#include "funcapi.h"
+#include "miscadmin.h"
+#include "postmaster/postmaster.h"
+#include "storage/lmgr.h"
+#include "utils/builtins.h"
+#include "utils/fmgroids.h"
+#include "utils/json.h"
+#include "utils/lsyscache.h"
+#include "utils/memutils.h"
+#include "utils/pg_lsn.h"
+#include "utils/syscache.h"
+#include "common/hashfn.h"
+#include "utils/varlena.h"
+#include "utils/guc_tables.h"
+#include "distributed/commands/utility_hook.h"
+
+
+#define MAX_SHARD_SIZE  20000
+PG_FUNCTION_INFO_V1(citus_auto_shard_split_start); 
+
+StringInfo GetSplitQuery(int64 shardminvalue , int64 shardmaxvalue , int64 nodeid , int64 shardid){
+
+    int64 midpoint = shardminvalue + ((shardmaxvalue - shardminvalue) >> 1);
+    StringInfo query = makeStringInfo();
+    appendStringInfo(query,"SELECT citus_split_shard_by_split_points(%ld, ARRAY['%ld'], ARRAY[%ld, %ld], 'block_writes')",
+        shardid, midpoint, nodeid, nodeid);      
+    return query;
+
+}
+
+
+void CitusGetShardData(){
+    
+
+    StringInfo query = makeStringInfo();
+    bool isnull;
+
+    /* This query is written to group the shards on the basis of colocation id and shardminvalue and get the shards with maximum shardsize
+     which is greater than a threshold . So for that first pg_dist_shard and citus_shards are joined followed by the joining of pg_dist_node
+     and citus_shards and finally joined by the table obtained by the grouping of colocation id and shardminvalue and shardsize exceeding the threshold.*/
+
+    appendStringInfoString(
+		query,
+		" SELECT cs.shardid,pd.shardminvalue,pd.shardmaxvalue,cs.shard_size,pn.nodeid"
+        " FROM pg_catalog.pg_dist_shard pd JOIN pg_catalog.citus_shards cs ON pd.shardid = cs.shardid JOIN pg_catalog.pg_dist_node pn ON cs.nodename = pn.nodename AND cs.nodeport = pn.nodeport"
+        " JOIN ("
+        " SELECT cs.colocation_id,pd.shardminvalue,MAX(cs.shard_size) AS max_shard_size"
+        " FROM pg_catalog.citus_shards cs JOIN pg_catalog.pg_dist_shard pd ON cs.shardid = pd.shardid"
+        " GROUP BY cs.colocation_id,pd.shardminvalue"
+        " HAVING MAX(cs.shard_size) > 20000"
+        " )AS max_sizes ON cs.colocation_id = max_sizes.colocation_id AND pd.shardminvalue = max_sizes.shardminvalue AND cs.shard_size = max_sizes.max_shard_size"
```

Comment:
can the group by and having move to the outer query?

(also max->sum)

Reply:
Changed the query